### PR TITLE
fst: init at 0.4.5

### PR DESCRIPTION
**Pull request by rmcgibbo**
###### Motivation for this change
Adds [this package](https://github.com/BurntSushi/fst). Mostly, this is a rust library, but it also includes a nice binary. The binary lets you pack a very large map or a set into a file and then search it quite fast using fuzzy string search, which can otherwise be fairly slow.

One question mark for me is that the upstream does not include a Cargo.lock file, so I had to make my own and include it as a patch. I'm not an expert on rust, so I don't know if that is the proper way for nixpkgs.

###### Things done

- [X] Tested using sandboxing ([nix.useSandbox](https://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](https://nixos.org/nix/manual/#sec-conf-file) on non-NixOS linux)
- Built on platform(s)
   - [X] NixOS
   - [ ] macOS
   - [ ] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [X] Tested compilation of all pkgs that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Ensured that relevant documentation is up to date
- [X] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

### CODE REVIEW

File: pkgs/top-level/all-packages.nix [hunk 1]
```diff
@@ -2337,6 +2337,8 @@ in
 
   fsmon = callPackage ../tools/misc/fsmon { };
 
+  fst = callPackage ../tools/text/fst {};
```
Comment (SuperSandro2000) [inline pkgs/top-level/all-packages.nix:2340]:
```suggestion
  fst = callPackage ../tools/text/fst { };
```

**Comment by SuperSandro2000** [comment]
> One question mark for me is that the upstream does not include a Cargo.lock file, so I had to make my own and include it as a patch. I'm not an expert on rust, so I don't know if that is the proper way for nixpkgs.

Exactly what we do.

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 115214` run on x86_64-linux [1](https://github.com/Mic92/nixpkgs-review)
<details>
<summary>
1 package built:
</summary>
<ul>
<li>fst</li>
</ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
fst:
<details>


warning: unused-argument
Unused argument: `stdenv`.
Near pkgs/tools/text/fst/default.nix:2:3:

```
  |
2 | , stdenv
  |   ^
```
</details>
</details>

File: pkgs/tools/text/fst/default.nix [hunk 2]
```diff
@@ -0,0 +1,42 @@
+{ lib
+, stdenv
```
Comment (SuperSandro2000) [inline pkgs/tools/text/fst/default.nix:2]:
```suggestion
```

**Comment by SuperSandro2000** [comment]
This is a semi-automatic executed nixpkgs-review with [nixpkgs-review-checks extension](https://github.com/SuperSandro2000/nixpkgs-review-checks). It is checked by a human on a best effort basis and does not build all packages (e.g. lumo, tensorflow or pytorch).
If you have any questions or problems please reach out to SuperSandro2000 on IRC.

Result of `nixpkgs-review pr 115214` run on x86_64-darwin [1](https://github.com/Mic92/nixpkgs-review)
<details>

  <summary>1 package failed to build and already failed to build on hydra master:</summary>
  <ul>
<li>fst: <a href="https://termbin.com/nsmv">plain log</a>  |  <a href="https://htmlpreview.github.io/?https://gist.githubusercontent.com/SuperSandroBot/7a8f33bf487f495f69c67d6f3292700c/raw/index.html">pretty log </a></li>
  </ul>
</details>

The following issues got detected with the above build packages.
Please fix at least the ones listed with your changed packages:

<details>
fst:
<details>


warning: unused-argument
Unused argument: `stdenv`.
Near pkgs/tools/text/fst/default.nix:2:3:

```
  |
2 | , stdenv
  |   ^
```
</details>
</details>

darwin patch which needs stdenv as an input.
I thought I could get away with a simple suggestion and squash. :D

```
diff --git a/pkgs/tools/text/fst/default.nix b/pkgs/tools/text/fst/default.nix
index 1d2dfd3ecd1..738fd69dc64 100644
--- a/pkgs/tools/text/fst/default.nix
+++ b/pkgs/tools/text/fst/default.nix
@@ -2,6 +2,7 @@
 , stdenv
 , fetchFromGitHub
 , rustPlatform
+, libiconv
 }:
 
 rustPlatform.buildRustPackage rec {
@@ -23,6 +24,8 @@ rustPlatform.buildRustPackage rec {
   cargoBuildFlags = [ "--workspace" ];
   cargoSha256 = "0svn2gzipslz939396rcydqx3i1x07l7acas7fhql12n59n2yrxw";
 
+  buildInputs = lib.optionals stdenv.isDarwin [ libiconv ];
+
   doInstallCheck = true;
   installCheckPhase = ''
     csv="$(mktemp)"
```

**Comment by rmcgibbo** [comment]
Thanks!

**Pull request merged**